pkg/compute/executor.go

```diff
@@ -41,13 +47,75 @@ func NewBaseExecutor(params BaseExecutorParams) *BaseExecutor {
 		ID:               params.ID,
 		callback:         params.Callback,
 		store:            params.Store,
+		Storages:         params.Storages,
 		executors:        params.Executors,
 		publishers:       params.Publishers,
 		failureInjection: params.FailureInjectionConfig,
 		resultsPath:      params.ResultsPath,
 	}
 }
 
+func PrepareRunArguments(ctx context.Context, strgprovider storage.StorageProvider, execution store.Execution, resultsDir string, cleanup *system.CleanupManager) (*executor.RunCommandRequest, error) {
+	inputVolumes, err := storage.ParallelPrepareStorage(ctx, strgprovider, execution.Job.Spec.Inputs...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepate storage for execution: %w", err)
+	}
+
+	cleanup.RegisterCallbackWithContext(func(ctx context.Context) error {
+		if err := storage.ParallelCleanStorage(ctx, strgprovider, inputVolumes); err != nil {
+			return fmt.Errorf("cleaning up job inputs: %w", err)
+		}
+		return nil
+	})
+
+	var engineArgs interface{}
+	engineArgs = execution.Job.Spec.Docker
+	if execution.Job.Spec.Engine == model.EngineWasm {
+		importModuleVolumes, err := storage.ParallelPrepareStorage(ctx, strgprovider, execution.Job.Spec.Wasm.ImportModules...)
+		if err != nil {
+			return nil, err
+		}
+		cleanup.RegisterCallbackWithContext(func(ctx context.Context) error {
+			if err := storage.ParallelCleanStorage(ctx, strgprovider, importModuleVolumes); err != nil {
+				return fmt.Errorf("cleaning up wasm import modules: %w", err)
+			}
+			return nil
+		})
+
+		entryModuleVolumes, err := storage.ParallelPrepareStorage(ctx, strgprovider, execution.Job.Spec.Wasm.EntryModule)
+		if err != nil {
+			return nil, err
+		}
+		cleanup.RegisterCallbackWithContext(func(ctx context.Context) error {
+			if err := storage.ParallelCleanStorage(ctx, strgprovider, entryModuleVolumes); err != nil {
+				return fmt.Errorf("cleaning up wasm entry modules: %w", err)
+			}
+			return nil
+		})
+		engineArgs = &wasm.Arguments{
+			EntryPoint:           execution.Job.Spec.Wasm.EntryPoint,
+			Parameters:           execution.Job.Spec.Wasm.Parameters,
+			EnvironmentVariables: execution.Job.Spec.Wasm.EnvironmentVariables,
+			EntryModule:          entryModuleVolumes[0],
+			ImportModules:        importModuleVolumes,
+		}
+	}
```

Comment:
I hate special casing the wasm engine, but for the time being this is workable since wasm is a "builtin" executor. This change fetches wasm modules on the compute node (from either IPFS or a URL) and makes them available locally before running the wasm executor. Previously the wasm executor was making network calls to compile the wasm binary before executing it in process.